File: 63-unique-paths-ii/unique-paths-ii.cpp
class Solution {
public:
    int solve(vector<vector<int>> &obstacleGrid, int m, int n, vector<vector<int>> &memo){
        if (m == 0 && n == 0){
            if (obstacleGrid[m][n] == 0) return 1;
            else return 0;
        }
        if (m < 0 || n < 0) return 0;
        if (obstacleGrid[m][n] == 1) return 0;
        if (memo[m][n] != -1) return memo[m][n];

        return memo[m][n] = solve(obstacleGrid, m-1, n, memo) + solve(obstacleGrid, m, n-1, memo);
    }
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        int m = obstacleGrid.size();
        int n = obstacleGrid[0].size();
        vector<vector<int>> dp(m+1, vector<int> (n+1, -1));
        return solve(obstacleGrid, m-1, n-1, dp);
    }
};

File: 63-unique-paths-ii/unique-paths-ii.go
func uniquePathsWithObstacles(obstacleGrid [][]int) int {
    m := len(obstacleGrid)
    n := len(obstacleGrid[0])

    dp := make([]int, n)
    flag := false
    for i := 0; i < n; i++ {
        if obstacleGrid[0][i] == 1 {
            flag = true
        }
        if flag {
            dp[i] = 0
        } else {
            dp[i] = 1
        }
    }

    for i := 2; i <= m; i++ {
        tmp := make([]int, n)
        for j := 0; j < n; j++ {
            up := dp[j]
            left := 0
            if j > 0 {
                left = tmp[j-1]
            }

            if obstacleGrid[i-1][j] == 1 {
                tmp[j] = 0
            } else {
                tmp[j] = up + left
            }
        }
        dp = tmp
    }
    return dp[n-1]
}